# iMAP Required Materials

<br>

<table class="table table-condensed" ,="" border="1">
<colgroup>
<col width="23%">
<col width="47%">
<col width="14%">
<col width="14%">
</colgroup>
<thead>
<tr class="header">
<th align="left"><strong>Requirement</strong></th>
<th align="left"><strong>Description</strong></th>
<th align="center"><strong>Location</strong></th>
<th align="center"><strong>Remarks</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left"><strong>Raw data</strong></td>
<td align="left">Demultiplexed reads in FASTQ format (.gz) with primers and barcodes removed</td>
<td align="center">data/raw</td>
<td align="center">fastq.gz</td>
</tr>
<tr class="even">
<td align="left"><strong>Sample metadata</strong></td>
<td align="left">File name: samplemetadata.tsv. A tab-separated file linking sample identifiers to the variables</td>
<td align="center">data/metadata</td>
<td align="center">Format: mothur or QIIME2</td>
</tr>
<tr class="odd">
<td align="left"><strong>Mapping files</strong></td>
<td align="left">For linking sample IDs to the data files</td>
<td align="center">data/metadata</td>
<td align="center">Mothur-formatted &amp; QIIME2-formatted</td>
</tr>
<tr class="even">
<td align="left"><em>Docker</em></td>
<td align="left">For creating Docker containers that wrap up iMAP dependencies.</td>
<td align="center">Docker Community Edition (CE)</td>
<td align="center"><a href="https://docs.docker.com/v17.12/install">Link</a></td>
</tr>
<tr class="odd">
<td align="left"><em>Seqkit</em></td>
<td align="left">For inspecting rawdata format and simple statistics.</td>
<td align="center">docker images: readqctools</td>
<td align="center"><a href="https://hub.docker.com/r/tmbuza/readqctools">Link</a></td>
</tr>
<tr class="even">
<td align="left"><em>BBduk.sh via BBMap</em></td>
<td align="left">For trimming poor quality reads and removing phiX contamination</td>
<td align="center">Auto-loaded at preprocessing step</td>
<td align="center"><a href="https://sourceforge.net/projects/bbmap/files/">Link</a></td>
</tr>
<tr class="odd">
<td align="left"><em>MultiQC</em></td>
<td align="left">For summarizing FASTQc output</td>
<td align="center">docker images: readqctools</td>
<td align="center"><a href="https://hub.docker.com/r/tmbuza/readqctools">Link</a></td>
</tr>
<tr class="even">
<td align="left"><em>Mothur</em></td>
<td align="left">For sequence processing, taxonomy assignment and preliminary analysis</td>
<td align="center">docker images: mothur:v1.41.3</td>
<td align="center"><a href="https://cloud.docker.com/repository/docker/tmbuza/mothur">Link</a></td>
</tr>
<tr class="odd">
<td align="left"><em>QIIME2</em></td>
<td align="left">For sequence processing, taxonomy assignment and preliminary analysis</td>
<td align="center">docker images: qiime2core:v2020.2</td>
<td align="center"><a href="https://hub.docker.com/r/tmbuza/qiime2core:v2020.2">Link</a></td>
</tr>
<tr class="even">
<td align="left"><em>R</em></td>
<td align="left">For statistical analysis and visualization</td>
<td align="center">docker image:rpackages:v3.5.2</td>
<td align="center"><a href="https://cloud.docker.com/repository/docker/tmbuza/rpackages">Link</a></td>
</tr>
<tr class="odd">
<td align="left"><em>iTOL</em></td>
<td align="left">For displaying, annotating and managing phylogenetic trees</td>
<td align="center">Onlline</td>
<td align="center"><a href="http://itol.embl.de/">Link</a></td>
</tr>
<tr class="even">
<td align="left"><em>SILVA NR</em> (mothur)</td>
<td align="left">Mothur-formatted rRNA alignments</td>
<td align="center">data/references</td>
<td align="center"><a href="https://www.mothur.org/w/images/3/32/">Link</a></td>
</tr>
<tr class="odd">
<td align="left"><em>SILVA NR</em> (QIIME2)</td>
<td align="left">QIIME2-formatted classifiers 515-806 fragments</td>
<td align="center">data/references</td>
<td align="center"><a href="https://data.qiime2.org/2020.2/common/silva-132-99-515-806-nb-classifier.qza">Link</a></td>
</tr>
<tr class="even">
<td align="left"><em>SILVA NR</em> (QIIME2)</td>
<td align="left">QIIME2-formatted classifiers full length</td>
<td align="center">data/references</td>
<td align="center"><a href="https://data.qiime2.org/2020.2/common/silva-132-99-nb-classifier.qza">Link</a></td>
</tr>
<tr class="odd">
<td align="left"><em>SILVA</em> (seed)</td>
<td align="left">Mothur-formatted rRNA alignments</td>
<td align="center">data/references</td>
<td align="center"><a href="https://www.mothur.org/w/images/7/71/">Link</a></td>
</tr>
<tr class="even">
<td align="left"><em>SILVA</em>(de-gapped)</td>
<td align="left">mothur-formatted classifiers</td>
<td align="center">data/references</td>
<td align="center">Auto-Generated</td>
</tr>
<tr class="odd">
<td align="left"><em>RDP</em></td>
<td align="left">Mothur-formatted classifiers</td>
<td align="center">data/references</td>
<td align="center"><a href="https://www.Mothur.org/wiki/RDP_reference_files">Link</a></td>
</tr>
<tr class="even">
<td align="left"><em>Greengenes</em></td>
<td align="left">Mothur-formatted classifiers</td>
<td align="center">data/references</td>
<td align="center"><a href="https://www.Mothur.org/wiki/Greengenes-formatted_databases">Link</a></td>
</tr>
<tr class="odd">
<td align="left"><em>Greengenes</em></td>
<td align="left">QIIME2-formatted classifiers 515-806 fragments</td>
<td align="center">data/references</td>
<td align="center"><a href="https://data.qiime2.org/2020.2/common/gg-13-8-99-515-806-nb-classifier.qza">Link</a></td>
</tr>
<tr class="even">
<td align="left"><em>Greengenes</em></td>
<td align="left">QIIME2-formatted classifiers full length</td>
<td align="center">data/references</td>
<td align="center"><a href="https://data.qiime2.org/2020.2/common/gg-13-8-99-nb-classifier.qza">Link</a></td>
</tr>
<tr class="odd">
<td align="left"><em>EzBioCloud</em></td>
<td align="left">Mothur-formatted classifiers</td>
<td align="center">data/references</td>
<td align="center"><a href="https://www.ezbiocloud.net/resources">Link</a></td>
</tr>
<tr class="even">
<td align="left"><em>Custom classifiesr</em></td>
<td align="left">Any manually built classifiers. Highly recommended when studying a specific group of known microbes.</td>
<td align="center">data/references</td>
<td align="center">Manually-built</td>
</tr>
</tbody>
</table>